homeassistant/components/mqtt/__init__.py

```diff
@@ -673,8 +721,9 @@ def async_disconnect(self):
         """
         def stop():
             """Stop the MQTT client."""
+            if self._task_reconnect:
+                self._task_reconnect.cancel()
```

Comment:
Asyncio is not threadsafe. You should not interact with it outside the loop and this function looks like in run in the thread of mqtt.

Reply:
Before changing, it was indeed run in the thread of mqtt/executor.
Now it's supposed to run in the main thread, isn't that the case when kicked off with:
```python
self._task_reconnect = \
            self.hass.async_create_task(self._mqtt_reconnect(result_code))
```

The only thing not running in main thread now should be this:
https://github.com/home-assistant/home-assistant/blob/0deb882ef31a8ac32a25a95adc6ab60d71dd4bc9/homeassistant/components/mqtt/__init__.py#L704-L705

Reply:
We need to review all the `async_add_job` or `add_job` calls.

The `stop` function here is scheduled on the executor since it isn't marked with callback. If we don't want to schedule `disconnect` on the executor we should change this.

Reply:
We should now be able to do both publish and subscribe all the way in the event loop.

Reply:
OK, right.

We just have to be a bit careful to not end up making calls back into the paho client from one of its callbacks as it will deadlock.
Maybe we can discuss on discord during next week?